### ISSUE TITLE
about train loss

### ISSUE THREAD

**Issue by zhangzhiyuan303** [Open-Superintelligence-Lab/deepseek-sparse-attention-research#7]
I'm trying to implement Indexer dense training as described in the paper, but it's not converging. Could you provide a demo? I saw that your code contains“compute_indexer_alignment_loss” 
thank you very much
**run_experiment_index.py**
def train_model(model, train_loader, val_loader, config, vocab_size, is_sparse=False):
    """Train a model"""
    indexer= model.blocks[0].attention.indexer
    optimizer = torch.optim.Adam(model.blocks[0].attention.indexer.parameters(), lr=config['learning_rate'])
    embed = nn.Embedding(vocab_size,256).to(config['device'])# 49152
    
    model.train()
    step = 0    
    while step < config['steps']:
        for batch in train_loader:
            if step >= config['steps']:
                break
            
            step_start = time.time()            
            input_ids, targets = batch
            input_ids = input_ids.to(config['device'])
            targets = targets.to(config['device'])
            x =embed(input_ids)
            output = indexer(x)
            attn = torch.einsum("bsd, btd->bst", x, x)
            attn_scores =F.softmax(attn)

            loss = F.kl_div(output.softmax(-1).log(), attn_scores, reduction='batchmean',log_target=False)            
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

**Comment by zhangzhiyuan303** [comment]
thank you

**Comment by vukrosic** [comment]
@zhangzhiyuan303 are you still working on this? I don't have time to work for it now, what are you trying to do, come to our discord and discuss - https://discord.gg/6AbXGpKTwN
